### PR TITLE
DROOLS-2301: Inline controls: Decision Table editor

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.jboss.org/browse/DROOLS-2301

This moves all Decision Table functions to be "inline". See [here](https://youtu.be/UaeUBln93JU) for a video.

### CODE REVIEW

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/dtable/RowNumberColumnHeaderMetaData.java [hunk 1]
```diff
@@ -51,25 +60,22 @@ public String getTitle() {
...
-                case MAX:
-                    sb.append(">");
-                    break;
-                case MIN:
-                    sb.append("<");
-                    break;
-                case SUM:
-                    sb.append("+");
+            if (builtinAggregator != null) {
+                switch (builtinAggregator) {
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/dtable/RowNumberColumnHeaderMetaData.java]: I think we could shorten the code if we store these aggregator characters directly in `BuiltinAggregator` enum.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/commands/expressions/types/dtable/SetHitPolicyCommandTest.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+import static org.mockito.Mockito.verifyZeroInteractions;
+
+@RunWith(MockitoJUnitRunner.class)
+public class SetHitPolicyCommandTest {
+
+    private static final HitPolicy OLD_HIT_POLICY = HitPolicy.ANY;
+
+    private static final HitPolicy NEW_HIT_POLICY = HitPolicy.UNIQUE;
+
+    private static final HitPolicy NEW_HIT_POLICY_COLLECT = HitPolicy.COLLECT;
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/commands/expressions/types/dtable/SetHitPolicyCommandTest.java]: This field is not used, so please introduce intended test.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/commands/expressions/types/dtable/AddInputClauseCommand.java [hunk 3]
```diff
@@ -73,12 +76,12 @@ public AddInputClauseCommand(final DecisionTable dtable,
 
             @Override
             public CommandResult<RuleViolation> execute(final GraphCommandExecutionContext context) {
-                dtable.getInput().add(inputClause);
-
+                final int icIndex = uiColumnIndex - DecisionTableUIModelMapperHelper.ROW_INDEX_COLUMN_COUNT;
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/commands/expressions/types/dtable/AddInputClauseCommand.java]: Changed (in new commit to follow). 

I also made naming in ```DeleteInputClauseCommand``` the same.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/commands/expressions/types/dtable/AddOutputClauseCommand.java [hunk 4]
```diff
@@ -73,12 +76,13 @@ public AddOutputClauseCommand(final DecisionTable dtable,
 
             @Override
             public CommandResult<RuleViolation> execute(final GraphCommandExecutionContext context) {
-                dtable.getOutput().add(outputClause);
+                final int ocIndex = uiColumnIndex - DecisionTableUIModelMapperHelper.ROW_INDEX_COLUMN_COUNT - dtable.getInput().size();
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/commands/expressions/types/dtable/AddOutputClauseCommand.java]: Changed (in new commit to follow). 

I also made naming in ```DeleteOutputClauseCommand``` the same.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/commands/expressions/types/dtable/AddInputClauseCommand.java [hunk 5]
```diff
@@ -46,19 +46,22 @@
     private final InputClause inputClause;
     private final GridData uiModel;
     private final InputClauseColumn uiModelColumn;
+    private final int uiColumnIndex;
     private final DecisionTableUIModelMapper uiModelMapper;
     private final org.uberfire.mvp.Command canvasOperation;
 
     public AddInputClauseCommand(final DecisionTable dtable,
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/commands/expressions/types/dtable/AddInputClauseCommand.java]: Good catch. Added and updated *all* Add/Delete Rule/Input/Output tests (in new commit to follow).

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/commands/expressions/types/dtable/AddOutputClauseCommand.java [hunk 6]
```diff
@@ -46,19 +46,22 @@
     private final OutputClause outputClause;
     private final GridData uiModel;
     private final OutputClauseColumn uiModelColumn;
+    private final int uiColumnIndex;
     private final DecisionTableUIModelMapper uiModelMapper;
     private final org.uberfire.mvp.Command canvasOperation;
 
     public AddOutputClauseCommand(final DecisionTable dtable,
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/commands/expressions/types/dtable/AddOutputClauseCommand.java]: Good catch. Added.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/commands/expressions/types/dtable/SetBuiltInAggregatorCommand.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+import org.kie.workbench.common.stunner.core.client.command.CanvasCommandResultBuilder;
+import org.kie.workbench.common.stunner.core.client.command.CanvasViolation;
+import org.kie.workbench.common.stunner.core.command.Command;
+import org.kie.workbench.common.stunner.core.command.CommandResult;
+import org.kie.workbench.common.stunner.core.graph.command.GraphCommandExecutionContext;
+import org.kie.workbench.common.stunner.core.graph.command.GraphCommandResultBuilder;
+import org.kie.workbench.common.stunner.core.graph.command.impl.AbstractGraphCommand;
+import org.kie.workbench.common.stunner.core.rule.RuleViolation;
+
+public class SetBuiltInAggregatorCommand extends AbstractCanvasGraphCommand implements VetoExecutionCommand,
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/commands/expressions/types/dtable/SetBuiltInAggregatorCommand.java]: Done. The underlying DMN model uses "Built*i*n" so I've done the same.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/commands/expressions/types/dtable/SetHitPolicyCommand.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+            @Override
+            public CommandResult<RuleViolation> execute(final GraphCommandExecutionContext gce) {
+                dtable.setHitPolicy(hitPolicy);
+
+                return GraphCommandResultBuilder.SUCCESS;
+            }
+
+            @Override
+            public CommandResult<RuleViolation> undo(final GraphCommandExecutionContext gce) {
+                dtable.setHitPolicy(oldHitPolicy.orElse(HitPolicy.UNIQUE));
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/commands/expressions/types/dtable/SetHitPolicyCommand.java]: The underlying model definition uses ```UNIQUE``` as the default it ```null```.

See https://github.com/kiegroup/drools/blob/master/kie-dmn/kie-dmn-model/src/main/java/org/kie/dmn/model/v1_1/DecisionTable.java#L54.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/commands/expressions/types/dtable/SetHitPolicyCommandTest.java (hunk 2 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/commands/expressions/types/dtable/SetHitPolicyCommandTest.java]: I've delete the constant.

The test for ```COLLECT``` is here: https://github.com/kiegroup/kie-wb-common/pull/1505/files#diff-314840c529071ca4e2761270ba248d75R598

Originally I was to have the ```SetHitPolicyCommand``` set the ```BuiltinAggregator``` if the HP was ```COLLECT``` but I ended up using a composite command from ```DecisionTableGrid```.

**Comment by manstis** [comment]
@jomarko All "first part" comments addressed in the second commit. Thank-you.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/commands/expressions/types/dtable/AddDecisionRuleCommandTest.java [hunk 9]
```diff
@@ -234,5 +299,66 @@ public void testCanvasCommandAddRuleAndThenUndo() throws Exception {
...
+        final Command<AbstractCanvasHandler, CanvasViolation> canvasCommand = command.newCanvasCommand(canvasHandler);
+        graphCommand.execute(graphCommandExecutionContext);
+        canvasCommand.execute(canvasHandler);
+
+        assertEquals(2,
+                     uiModel.getRowCount());
+        assertEquals(existingUiRow,
+                     uiModel.getRow(0));
+        assertEquals(uiModelRow,
+                     uiModel.getRow(1));
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/commands/expressions/types/dtable/AddDecisionRuleCommandTest.java]: We could check also cell values of `uiModel.getRow()`.

Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/commands/expressions/types/dtable/AddDecisionRuleCommandTest.java]: Done in fourth commit.

**Pull request merged**